library/core/src/primitive_docs.rs

```diff
@@ -1506,6 +1506,113 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer with
+/// a different signature, the two signatures must be *ABI-compatible* or else calling the function
+/// via that function pointer is Undefined Behavior. ABI compatibility is a lot stricter than merely
+/// having the same memory layout; for example, even if `i32` and `f32` have the same size and
+/// alignment, they might be passed in different registers and hence not be ABI-compatible.
+///
+/// ABI compatibility as a concern only arises in code that alters the type of function pointers,
+/// code that imports functions via `extern` blocks, and in code that combines `#[target_feature]`
+/// with `extern fn`. Altering the type of function pointers is wildly unsafe (as in, a lot more
+/// unsafe than even [`transmute_copy`][mem::transmute_copy]), and should only occur in the most
+/// exceptional circumstances. Most Rust code just imports functions via `use`. `#[target_feature]`
+/// is also used rarely. So, most likely you do not have to worry about ABI compatibility.
+///
+/// But assuming such circumstances, what are the rules? For this section, we are specifically
+/// concerned with the case where both the caller and the callee are defined in Rust.
```

Comment:
```suggestion
/// concerned with the case where both the caller and the callee are defined in Rust (and using the same compiler version).
```

Right? We're not making guarantees about inter-op between future and past compiler versions here, in terms of what the ABI of some primitive is?

Reply:
If you start linking separately compiled Rust objects with each other there's a *ton* of things you need to consider. Using the same Rust version is necessary but far from sufficient. I don't even know the full list of things, but there's a bunch of flags that must be set the same, as far as I understand.

I'm not sure if we want to get into listing all these requirements here. I want to discuss ABI, not linking